### PR TITLE
Implement Trilinos AztecOO_StatusTest for ReductionControl.

### LINKED ISSUE #3843: ReductionControl and the Trilinos solvers
I am trying to debug `lac/schur_complement_05.debug`, which fails with a floating point exception. The cause is that we initialize the `ReductionControl` object in the following way:
```
ReductionControl::ReductionControl(const unsigned int n,
                                   const double tol,
                                   const double red,
                                   const bool m_log_history,
                                   const bool m_log_result)
  :
  SolverControl (n, tol, m_log_history, m_log_result),
  reduce(red),
  reduced_tol(numbers::signaling_nan<double>())
{}
```
The point of the last line is that we must not read `reduced_tol` unless it is set in the first iteration. This happens as follows:
```
SolverControl::State
ReductionControl::check (const unsigned int step,
                         const double check_value)
{
  // if this is the first time we
  // come here, then store the
  // residual for later comparisons
  if (step==0)
    {
      initial_val = check_value;
      reduced_tol = check_value * reduce;
    }

  // check whether desired reduction
  // has been achieved. also check
  // for equality in case initial
  // residual already was zero
  if (check_value <= reduced_tol)
    {
      ...
```
In other words, in the first iteration we get the first residual (`check_value`) and multiply it by the desired reduction, and then `reduced_tol` contains a usable value.

The problem in this test is that apparently `ReductionControl::check()` is not, in fact, called in each iteration. The test uses the `TrilinosWrappers::CG` as solver, and at least as far as I can see, the first time `check()` is called is when `step` equals 74. This is because our Trilinos solver wrappers seem to pretty explicitly work around the design of `SolverControl`/`ReductionControl`:
```
  template<typename Preconditioner>
  void
  SolverBase::do_solve(const Preconditioner &preconditioner)
  {
    int ierr;

    // Next we can allocate the AztecOO solver...
    solver.SetProblem(*linear_problem);

    // ... and we can specify the solver to be used.
    switch (solver_name)
      {
      case cg:
        solver.SetAztecOption(AZ_solver, AZ_cg);
        break;
      case cgs:
        solver.SetAztecOption(AZ_solver, AZ_cgs);
        break;
      case gmres:
        solver.SetAztecOption(AZ_solver, AZ_gmres);
        solver.SetAztecOption(AZ_kspace, additional_data.gmres_restart_parameter);
        break;
      case bicgstab:
        solver.SetAztecOption(AZ_solver, AZ_bicgstab);
        break;
      case tfqmr:
        solver.SetAztecOption(AZ_solver, AZ_tfqmr);
        break;
      default:
        Assert (false, ExcNotImplemented());
      }

    // Set the preconditioner
    set_preconditioner(solver, preconditioner);

    // ... set some options, ...
    solver.SetAztecOption (AZ_output, additional_data.output_solver_details ?
                           AZ_all : AZ_none);
    solver.SetAztecOption (AZ_conv, AZ_noscaled);

    // ... and then solve!
    ierr = solver.Iterate (solver_control.max_steps(),        // *****************************
                           solver_control.tolerance());                   // *****************************

    // report errors in more detail than just by checking whether the return
    // status is zero or greater. the error strings are taken from the
    // implementation of the AztecOO::Iterate function
    switch (ierr)
      {
      case -1:
        AssertThrow (false, ExcMessage("AztecOO::Iterate error code -1: "
                                       "option not implemented"));
      case -2:
        AssertThrow (false, ExcMessage("AztecOO::Iterate error code -2: "
                                       "numerical breakdown"));
      case -3:
        AssertThrow (false, ExcMessage("AztecOO::Iterate error code -3: "
                                       "loss of precision"));
      case -4:
        AssertThrow (false, ExcMessage("AztecOO::Iterate error code -4: "
                                       "GMRES Hessenberg ill-conditioned"));
      default:
        AssertThrow (ierr >= 0, ExcTrilinosError(ierr));
      }

    // Finally, let the deal.II SolverControl object know what has
    // happened. If the solve succeeded, the status of the solver control will
    // turn into SolverControl::success.
    solver_control.check (solver.NumIters(), solver.TrueResidual());   // *****************************

    if (solver_control.last_check() != SolverControl::success)
      AssertThrow(false, SolverControl::NoConvergence (solver_control.last_step(),
                                                       solver_control.last_value()));
```

Note that the marked lines are the only interaction with the `SolverControl` class. In particular, we query `solver_control.tolerance()` before it is even set to anything useful.

Ideas on how to fix this?

### PR DESCRIPTION
**Pull request by jppelteret**
This patch ensures that Trilinos solvers now respect the convergence
criterion specified by ReductionControl. To do so, we create an
AztecOO_StatusTest that monitors the heuristics that ReductionControl
uses to test for convergence.

This hopefully corrects the issue in #3843, but we'll wait to see if it passes on the tester.

### CODE REVIEW

**Comment by davydden** [comment]
a unit which uses `ReductionControl` with trilinos and prints the number of iterations would be nice.

**Comment by jppelteret** [comment]
**TODO**:
- [ ] Work out why `ReductionControl` reports failure when the absolute tolerance is met (The linear solver correctly stops at this point, but `solver_control.last_check()` does not report success).
- [ ] Double check my additional comments in the code. Although `Trilinos`' default convergence check is analogous to our `IterationNumberControl`, `deal.II` ultimately still decides whether the solver has succeeded or failed. i.e. whether hitting the max number of steps constitutes failure (`SolverControl`) or success (`IterationNumberControl`)

File: source/lac/trilinos_solver.cc [hunk 1]
```diff
@@ -282,6 +291,143 @@ namespace TrilinosWrappers
...
+    namespace
+    {
+      double
+      compute_residual (const Epetra_MultiVector *const residual_vector)
+      {
+        Assert(residual_vector->NumVectors() == 1,
+               ExcMessage("Residual multivector holds more than one vector"));
+        std_cxx11::array<double, 1> res_l2_norm;
+        residual_vector->Norm2(res_l2_norm.data());
+        return res_l2_norm[0];
```
Comment (kronbichler) [inline source/lac/trilinos_solver.cc]: I think an `std::array` is a bit of an overkill here. We typically use a single variable `double res_l2_norm` and then pass in `&res_l2_norm` to `Norm2`. Have a look at the use of `Norm2` in `include/deal.II/lac/trilinos_vector.h`.

Comment (jppelteret) [inline source/lac/trilinos_solver.cc]: Yes, you're right. I was thinking too much about pointers!

**Comment by jppelteret** [comment]
Thanks a lot for looking over this @kronbichler! Yes, as best I can tell `ReductionControl` + `Trilinos` iterative solvers are now behaving as expected. The problem test referred to in #3843  (`lac/schur_complement_05.debug`) now converges in fewer linear iterations. From the logs it looks like it obeys the prescribed reduction as well. The test passes on my machine, so I hope that this fixes the issue seen there.

**Comment by jppelteret** [comment]
@bangerth If you've got no comments on this then I think that its ready to merge.

**Pull request merged**